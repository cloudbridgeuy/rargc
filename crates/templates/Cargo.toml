[package]
name = "templates"
version = "0.0.0"
homepage = "https://github.com/cloudbridgeuy/rargc"
description = "Template library for rargc"
autobins = false

authors.workspace = true
edition.workspace = true
license.workspace = true

[lib]
name = "templates"
path = "src/lib.rs"

[dependencies]
color-eyre = "0.6.2"
log = "0.4.19"
tera = "1"
lazy_static = "1.4.0"
serde = { version = "1.0.186", features = ["derive"] }
serde_json = "1.0.105"
include_dir = "0.7.3"

[dev-dependencies]
insta = { version = "1.31.0", features = ["yaml"] }



