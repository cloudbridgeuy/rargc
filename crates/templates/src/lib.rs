use std::collections::HashMap;

use color_eyre::eyre::Result;
use include_dir::{include_dir, Dir};
use lazy_static::lazy_static;
use serde::Serialize;
use serde_json::value::{Map, Value};
use tera::{Context, Tera};

static TEMPLATES_DIR: Dir = include_dir!("$CARGO_MANIFEST_DIR/templates");

/// Merges two Map<String, Value> objects. The second object takes precedence over the first.
pub fn merge_maps(
    first: &mut Map<String, Value>,
    second: &Map<String, Value>,
) -> Map<String, Value> {
    for (key, second_value) in second {
        if let Some(first_value) = first.get_mut(key) {
            if let (Some(first_value), Some(second_value)) =
                (first_value.as_object_mut(), second_value.as_object())
            {
                merge_maps(first_value, second_value);
            } else {
                *first_value = second_value.clone();
            }
        } else {
            first.insert(key.clone(), second_value.clone());
        }
    }

    first.clone()
}

/// Tera filter that uses the merge_json_objects function to merge two values that are JSON objects.
pub fn merge_json_objects_filter(
    value: &Value,
    args: &HashMap<String, Value>,
) -> tera::Result<Value> {
    let with = args.get("with").ok_or_else(|| {
        tera::Error::msg("The filter `merge_json_objects` requires an argument named `with`")
    })?;
    let with = with.as_object().ok_or_else(|| {
        tera::Error::msg(format!(
            "The filter `merge_json_objects` received an invalid argument: {:?}",
            args
        ))
    })?;
    let value = value.as_object().ok_or_else(|| {
        tera::Error::msg(format!(
            "The filter `merge_json_objects` received an invalid value: {:?}",
            value
        ))
    })?;

    let first = merge_maps(&mut value.clone(), with);
    Ok(serde_json::to_value(first).unwrap())
}

lazy_static! {
    pub static ref TEMPLATES: Tera = {
        let mut tera = match Tera::new("templates/**/*.tera") {
            Ok(t) => t,
            Err(e) => {
                log::error!("Parsing error(s): {}", e);
                ::std::process::exit(1);
            }
        };

        for entry in TEMPLATES_DIR.files() {
            match tera.add_raw_template(
                entry.path().to_str().unwrap(),
                entry.contents_utf8().unwrap(),
            ) {
                Ok(_) => (),
                Err(e) => {
                    log::error!("Parsing error(s): {}", e);
                    ::std::process::exit(1);
                }
            }
        }

        tera.register_filter("merge", merge_json_objects_filter);

        tera
    };
}

/// Render the main.sh template
pub fn render<T: Serialize>(context: &T) -> Result<String> {
    let output = TEMPLATES.render("main.tera", &Context::from_serialize(context)?)?;
    Ok(output)
}
