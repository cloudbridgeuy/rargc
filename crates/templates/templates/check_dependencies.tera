{%- if dep and dep | length > 0 -%}
  {%- for d in dep %}
  for dependency in {{ d["list"] | join(sep=" ")  }}; do
    if ! command -v $dependency >/dev/null 2>&1; then
      printf "missing dependency: $dependency\n" >&2
      printf "\e[31m%s\e[33m%s\e[31m\e[0m\n\n" "Missing dependency: " "$dependency" >&2
      {%- if dep["message"] %}
      printf "{{ dep["message"] }}\n" >&2
      {%- endif %}
      exit 1
    fi
  done
  {%- endfor %}
{%- endif %}
