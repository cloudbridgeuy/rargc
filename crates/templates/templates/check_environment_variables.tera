{%- set _envs = false %}
{%- if command["envs"] %}
  {%- if envs %}
    {%- set _envs = envs | merge(with=command["envs"]) %}
  {%- else %}
    {%- set _envs = command["envs"] %}
  {%- endif %}
{%- elif envs %}
  {%- set _envs = envs %}
{%- endif %}
{%- if _envs and _envs | length > 0 %}
  {%- for name, env in _envs %}
    {%- if env["option"] %}
  if [[ -n "${% raw %}{{% endraw %}{{ name }}:-}" ]]; then
    args['{{ env["option"] }}']="${% raw %}{{% endraw %}{{ name }}:-}"
  fi
    {%- elif env["required"] %}
  if [[ -z "${% raw %}{{% endraw %}{{ name }}:-}" ]]; then
    printf "\e[31m%s\e[33m%s\e[31m\e[0m\n\n" "Missing required environment variable: " "{{ name }}" >&2
      {%- if command %}
    {{ command["name"] }}_usage >&2
      {%- else %}
    usage >&2
      {%- endif %}
    exit 1
  fi
    {%- endif %}
  {%- endfor %}
{%- endif %}
