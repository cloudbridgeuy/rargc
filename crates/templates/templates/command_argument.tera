{% if command["short"] -%}-{{ command["short"] }} | {% endif -%}--{{ name }})
        action="{{ name }}"
        shift
        {{ name }}_parse_arguments "$@"
        shift $#
        ;;
