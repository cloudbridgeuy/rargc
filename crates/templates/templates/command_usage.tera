{%- set_global required = [] -%}
{%- if options -%}
  {%- for key, option in options -%}
    {%- if option["required"] -%}
      {%- set_global required = required | concat(with=option) -%}
    {%- endif -%}
  {%- endfor -%}
{%- endif -%}
{%- if command["options"] -%}
  {%- for key, option in command["options"] -%}
    {%- if option["required"] -%}
      {%- set_global required = required | concat(with=option) -%}
    {%- endif -%}
  {%- endfor -%}
{%- endif -%}

{{ command["name"] }}_usage() {
  {% if command["description"] -%}
  printf "{{ command["description"] }}\n"
  {% endif -%}

  {% if command["help"] -%}
  printf "{{ command["help"] }}\n"
  {% endif -%}

  printf "\n\033[4m%s\033[0m\n" "Usage:"
  {% if command["positional_arguments"] -%}
  printf "  {{ command["name"] }} {% for option in required -%}{% if option["short"] %}-{{
  option["short"]}}|{% endif %}--{{ option["name"] }} <{{ option["name"] | upper }}> {% endfor
%}[OPTIONS] {% for arg in command["positional_arguments"] %}{% if arg["required"] %}{% if arg["value_notation"] %}{{
arg["value_notation"] }}{% else %}{{ arg["name"] | upper }}{% endif %}{% if arg["multiple"] %}...{% endif %}{% else %}[{% if arg["value_notation"] %}{{
arg["value_notation"] }}{% else %}{{ arg["name"] | upper }}{% endif %}{% if arg["multiple"] %}...{% endif %}]{% endif %} {% endfor %}\n"
  {% else -%}
  printf "  {{ command["name"] }} {% for option in required %}{% if option["short"] %}-{{ option["short"]}}|{% endif %}--{{ option["name"] }} <{{ option["name"] | upper }}> {% endfor %}[OPTIONS]\n"
  {% endif -%}
  printf "  {{ command["name"] }} -h|--help\n"

  printf "\n\033[4m%s\033[0m\n" "Options:"
  {%- if command["options"] or command["flags"] %}
  {%- if command["options"] %}
  {%- for key, option in command["options"] %}
  printf "  {% if option["short"] -%}-{{ option["short"] }} {% endif -%}--{{ key }} {% if option["required"] %}<{{ key | upper }}>{% else %}[<{{ key | upper }}>]{% endif %}\n"
  printf "    {{ option["summary"] }}\n"
  {%- if option["multiple"] or option["default"] or option["choices"] %}
  printf "    [
    {%- if option["multiple"] -%}
      @multiple
    {%- endif -%}
    {%- if option["default"] -%}
      {%- if option["multiple"] -%}, {% endif -%}@default {{ option["default"] }}
    {%- endif -%}
    {%- if option["choices"] -%}
      {%- if option["multiple"] or option["default"] -%}, {% endif -%}@choices {{ option["choices"] | join(sep=", ") }}
    {%- endif -%}]\n"
  {%- endif -%}
  {%- endfor -%}
  {%- endif %}
  {%- endif %}
  printf "  -h --help\n"
  printf "    Print help\n"

  {%- if options or flags %}
  printf "\n\033[4m%s\033[0m\n" "Global Options:"
  {%- if options %}
  {%- for key, option in options %}
  printf "  {% if option["short"] -%}-{{ option["short"] }} {% endif -%}--{{ key }} {% if option["required"] %}<{{ key | upper }}>{% else %}[<{{ key | upper }}>]{% endif %}\n"
  printf "    {{ option["summary"] }}\n"
  {%- if option["multiple"] or option["default"] or option["choices"] %}
  printf "    [
    {%- if option["multiple"] -%}
      @multiple
    {%- endif -%}
    {%- if option["default"] -%}
      {%- if option["multiple"] -%}, {% endif -%}@default {{ option["default"] }}
    {%- endif -%}
    {%- if option["choices"] -%}
      {%- if option["multiple"] or option["default"] -%}, {% endif -%}@choices {{ option["choices"] | join(sep=", ") }}
    {%- endif -%}]\n"
  {%- endif -%}
  {%- endfor -%}
  {%- endif %}
  {%- if flags %}
  {%- for key, flag in flags %}
  printf "  {% if flag["short"] -%}-{{ flag["short"] }} {% endif -%}--{{ key }}\n"
  printf "    {{ flag["summary"] }}\n"
  {%- endfor %}
  {%- endif %}
  {%- endif %}
}
