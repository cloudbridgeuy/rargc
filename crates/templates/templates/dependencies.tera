# Check global dependencies
check_global_dependencies() {
