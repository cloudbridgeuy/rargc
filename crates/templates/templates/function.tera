{%- set_global required = [] -%}
{%- set_global defaults = [] -%}
{%- set_global choices = [] -%}
{%- set_global _any = false %}
{%- if options -%}
  {%- if command["options"] %}
    {%- for key, option in options | merge(with=command["options"]) -%}
      {%- if option["required"] -%}
        {%- set_global required = required | concat(with=option) -%}
      {%- endif -%}
      {%- if option["default"] -%}
        {%- set_global defaults = defaults | concat(with=option) -%}
      {%- endif -%}
      {% if option["choices"] -%}
        {%- set_global choices = choices | concat(with=option) -%}
      {%- endif -%}
    {%- endfor -%}
  {%- else %}
    {%- for key, option in options -%}
      {%- if option["required"] -%}
        {%- set_global required = required | concat(with=option) -%}
      {%- endif -%}
      {%- if option["default"] -%}
        {%- set_global defaults = defaults | concat(with=option) -%}
      {%- endif -%}
      {% if option["choices"] -%}
        {%- set_global choices = choices | concat(with=option) -%}
      {%- endif -%}
    {%- endfor -%}
  {%- endif %}
{%- else -%}
  {%- if command["options"] -%}
    {%- for key, option in command["options"] -%}
      {%- if option["required"] -%}
        {%- set_global required = required | concat(with=option) -%}
      {%- endif -%}
      {%- if option["default"] -%}
        {%- set_global defaults = defaults | concat(with=option) -%}
      {%- endif -%}
      {% if option["choices"] -%}
        {%- set_global choices = choices | concat(with=option) -%}
      {%- endif -%}
    {%- endfor -%}
  {%- endif -%}
{%- endif -%}
{%- if command["positional_arguments"] -%}
  {%- for arg in command["positional_arguments"] -%}
    {%- if arg["required"] -%}
      {%- set_global required = required | concat(with=arg) -%}
    {%- endif -%}
    {%- if arg["default"] -%}
      {%- set_global defaults = defaults | concat(with=arg) -%}
    {%- endif -%}
    {% if arg["choices"] -%}
      {%- set_global choices = choices | concat(with=arg) -%}
    {%- endif -%}
  {%- endfor -%}
{%- endif -%}
{%- set _envs = false %}
{%- if command["envs"] %}
  {%- if envs %}
    {%- set _envs = envs | merge(with=command["envs"]) %}
  {%- else %}
    {%- set _envs = command["envs"] %}
  {%- endif %}
{%- elif envs %}
  {%- set _envs = envs %}
{%- endif %}
{%- if any %}
  {%- if command["any"] %}
    {%- set _any = command["any"] %}
  {%- else %}
    {%- set _any = any %}
  {%- endif %}
{%- elif command["any"] %}
  {%- set _any = command["any"] %}
{%- endif %}
{%- set opts = false %}
{%- set fls = false %}
{%- set args = false %}
{%- if options %}
  {%- if command["options"] %}
    {%- set opts = options | merge(with=command["options"]) %}
  {%- else %}
    {%- set opts = options %}
  {%- endif %}
{%- else %}
  {%- if command["options"] %}
    {%- set opts = command["options"] %}
  {%- endif %}
{%- endif %}
{%- if flags %}
  {%- if command["flags"] %}
    {%- set fls = flags | merge(with=command["flags"]) %}
  {%- else %}
    {%- set fls = flags %}
  {%- endif %}
{%- else %}
  {%- if command["flags"] %}
    {%- set fls = command["flags"] %}
  {%- endif %}
{%- endif %}
{%- if positional_arguments %}
  {%- if command["positional_arguments"] %}
    {%- set args = command["positional_arguments"] %}
  {%- else %}
    {%- set args = positional_arguments %}
  {%- endif %}
{%- else %}
  {%- if command["positional_arguments"] %}
    {%- set args = command["positional_arguments"] %}
  {%- endif %}
{%- endif %}

{%- if command["description"] -%}
# {{ command["description"] }}
{% endif -%}
{% if command["help"] -%}
# {{ command["help"] }}
{% endif -%}
{{ command["name"] }}() {
{%- if opts %}
  {%- for name, option in opts %}
    {%- if option["multiple"] %}
  declare -a rargs_{{ name | replace(from="-", to="_") }}
    {%- endif %}
  {%- endfor %}
{%- endif %}

{%- if args %}
  {%- for arg in args %}
    {%- if arg["multiple"] %}
  declare -a rargs_{{ name | replace(from="-", to="_") }}
    {%- endif %}
  {%- endfor %}
{%- endif %}

{%- if _envs and _envs | length > 0 %}
  # Parse environment variables
  {% include "check_environment_variables.tera" %}
{%- endif %}
  # Parse command arguments
  parse_{{ command["name"] }}_arguments "$@"
{% if rules -%}
  {%- if rules is containing("no-first-option-help") and rules is not containing("custom-usage") %}
  # Rule `no-first-option-help`: Render the global or command usage if the `-h|--help` option is
  #                              is provided anywhere on the command, not just as the first option.
  #                              Handling individual functions case by case.
  if [[ -n "$rargs_help" ]]; then
    {{ command["name"] }}_usage
    exit 0
  fi
  {%- endif -%}
{%- endif -%}
{%- if command["dep"] and command["dep"] | length > 0 -%}
  {% set dep = command["dep"] %}
  # Check dependencies
  {%- include "check_dependencies.tera" %}
{%- endif %}
{%- if defaults | length > 0 %}
  {% for option in defaults %}
    {% if option["multiple"] %}
  if [[ "${% raw %}{{% endraw %}{% raw %}#{% endraw %}rargs_{{ option["name"] | replace(from="-", to="_") }}[@]}" -eq 0 ]]; then
    rargs_{{ option["name"] | replace(from="-", to="_") }}=({{ option["default"] }})
  fi
    {% else %}
  if [[ -z "$rargs_{{ option["name"] | replace(from="-", to="_") }}" ]]; then
    rargs_{{ option["name"] | replace(from="-", to="_") }}="{{ option["default"] }}"
  fi
    {% endif %}
  {%- endfor %}
{%- endif %}
{%- if choices | length > 0 %}
  {% for option in choices %}
  if [[ -n "$rargs_{{ option["name"] | replace(from="-", to="_") }}" ]]; then
    {%- if option["multiple"] %}
    for value in "${rargs_{{ option["name"] | replace(from="-", to="_") }}[@]}"; do
      if [[ ! "({{ option["choices"] | join(sep=" ") }})" =~ ${value} ]]; then
        printf "\e[31m%s\e[33m%s\e[31m%s\e[33m%s\e[31m\e[0m\n\n" "Invalid option for " "{{ option["name"] }}" ": " "${value}" >&2
        {{ command["name"] }}_usage >&2
        exit 1
      fi
    done
    {%- else %}
    if [[ ! "({{ option["choices"] | join(sep=" ") }})" =~ $rargs_{{ option["name"] | replace(from="-", to="_") }} ]]; then
      printf "\e[31m%s\e[33m%s\e[31m%s\e[33m%s\e[31m\e[0m\n\n" "Invalid option for " "{{ option["name"] }}" ": " "$rargs_{{ option["name"] | replace(from="-", to="_") }}" >&2
      {{ command["name"] }}_usage >&2
      exit 1
    fi
    {%- endif %}
  fi
  {%- endfor %}
{%- endif %}
{%- if _any and _any["required"] %}
  if [[ "{%- raw %}${{%- endraw %}#other_args[@]}" == "0" ]]; then
    printf "\e[31m%s\e[0m\n\n" "Missing required additional argument" >&2
    {{ command["name"] }}_usage >&2
    {%- raw %}
    exit 1
  fi
  {%- endraw %}
{%- endif %}
{%- if required | length > 0 %}
  {% for option in required %}
    {%- if option["multiple"] %}
  if [[ "{% raw %}${{%endraw %}#rargs_{{ option["name"] | replace(from="-", to="_") }}[@]}" == 0 ]]; then
    {%- else %}
  if [[ -z "$rargs_{{ option["name"] | replace(from="-", to="_") }}" ]]; then
    {%- endif %}
    printf "\e[31m%s\e[33m%s\e[31m\e[0m\n\n" "Missing required option: " "{{ option["name"] }}" >&2
    {{ command["name"] }}_usage >&2
    exit 1
  fi
  {%- endfor %}
{%- endif %}
{%- if command["subcommand"] %}
  local sub="{{ command["subcommand"] }}"
{%- endif %}
{%- if command["lines"] %}
  {%- if command["lines"] | length > 0 %}
    {%- for line in command["lines"] %}
{{ line }}
    {%- endfor %}
  {%- endif %}
{%- else %}
  echo "No implementation body found for function {{ command["name"] }}"
{%- endif %}
}
