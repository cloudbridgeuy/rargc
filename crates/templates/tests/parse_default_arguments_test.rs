use tera::Context;
use test_log::test;

use templates::TEMPLATES;

#[test]
fn test_render() {
    let objects = vec![
        Context::from_serialize(serde_json::json!({
            "noname": "foo".to_string(),
        })),
        Context::from_serialize(serde_json::json!({
            "name": "foo".to_string(),
        })),
        Context::from_serialize(serde_json::json!({
            "noname": "foo".to_string(),
            "rules": ["no-first-option-help"]
        })),
        Context::from_serialize(serde_json::json!({
            "name": "foo".to_string(),
            "rules": ["no-first-option-help"]
        })),
        Context::from_serialize(serde_json::json!({
            "noname": "foo".to_string(),
            "rules": ["custom-usage"]
        })),
        Context::from_serialize(serde_json::json!({
            "noname": "foo".to_string(),
            "rules": ["no-first-option-help", "custom-usage"]
        })),
        Context::from_serialize(serde_json::json!({
            "name": "foo".to_string(),
            "rules": ["no-first-option-help", "custom-usage"]
        })),
    ];

    for object in objects {
        let output = match TEMPLATES.render(
            "parse_default_arguments.tera",
            &object.expect("Can't create JSON object"),
        ) {
            Ok(o) => o,
            Err(e) => {
                log::error!("Parsing error(s): {}", e);
                ::std::process::exit(1);
            }
        };

        insta::assert_snapshot!(output)
    }
}
