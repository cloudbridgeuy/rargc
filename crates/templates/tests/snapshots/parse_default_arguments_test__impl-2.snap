---
source: crates/templates/tests/parse_default_arguments_test.rs
expression: output
---
  while [[ $# -gt 0 ]]; do
    case "${1:-}" in
      -h | --help)
        foo_usage
        exit
        ;;

      *)
        break
        ;;
    esac
  done


